docs/src/architecture/notes.md

```diff
@@ -1,73 +1,191 @@
+
+  ### Main script
+  The main part of the P2ID script checks if the executing account is the same as the account defined in the `NoteInputs`. The creator of the note defines the note script and the note inputs separately to ensure usage of the same standardized P2ID script regardless of the target account ID. That way, it is enough to check the script root (see above).
+
+  ```
+  # Pay-to-ID script: adds all assets from the note to the account, assuming ID of the account
+  # matches target account ID specified by the note inputs.
+  #
+  # Requires that the account exposes: miden::contracts::wallets::basic::receive_asset procedure.
+  #
+  # Inputs: [SCRIPT_ROOT]
+  # Outputs: []
+  #
+  # Note inputs are assumed to be as follows:
+  # - target_account_id is the ID of the account for which the note is intended.
+  #
+  # FAILS if:
+  # - Account does not expose miden::contracts::wallets::basic::receive_asset procedure.
+  # - Account ID of executing account is not equal to the Account ID specified via note inputs.
+  # - The same non-fungible asset already exists in the account.
+  # - Adding a fungible asset would result in amount overflow, i.e., the total amount would be
+  #   greater than 2^63.
+  begin
+      # drop the transaction script root
+      dropw
+      # => []
+      
+      # load the note inputs to memory starting at address 0
+      push.0 exec.note::get_inputs
+        # => [inputs_ptr]
+  
+      # read the target account id from the note inputs
+      mem_load
+      # => [target_account_id]
+  
+      exec.account::get_id
+      # => [account_id, target_account_id, ...]
+  
+      # ensure account_id = target_account_id, fails otherwise
+      assert_eq
+      # => [...]
+  
+      exec.add_note_assets_to_account
+      # => [...]
+  end
+  ```
+
+  Every note script starts with the note script root on top of the stack. After the `dropw`, the stack is cleared. Next, the script loads the note inputs by first getting the pointer to the memory address of the inputs `exec.note::get_inputs`. The `push.0` just ensures that the pointer overrides the newly inserted `0`. Then, `mem_load` loads a `Felt` from the specified memory address and puts it on top of the stack, in that cases the   `target_account_id` defined by the creator of the note. Now, the note invokes `get_id` from the account API using `exec.account::get_id` - which is   possible even in the note context. Because, there are two account IDs on top of the stack now, `assert_eq` fails if the two account IDs (target_account_id and executing_account_id) are not the same. That means, the script cannot be successfully executed if executed by any other account than the account specified by the note creator using the note inputs.
```

Comment:
I think the part about the pointer is a bit unclear here - especially the part about "The push.0 just ensures that the pointer overrides the newly inserted 0".

What is happening here is that `push.0` defines the value of `inputs_ptr` and then `exec.note::get_inputs` writes the inputs into memory starting at that location. We could have just as easily used `push.1` (or any other value) and written the inputs at that memory address.